boards: nrf9160dk: Add support for newer revisions (0.14.0+) of the DK

anangl opened the pull request:
- Use devicetree instead of Kconfig to configure the board control switches in nRF9160 DK:
  - add binding for the switches that provide optional signal routings on this board
  - add binding for the GPIO interface that can be used for communication (e.g. UART based) between the nRF9160 and the nRF52840 on the DK, and add GPIO mapping for this interface so that its lines can be used without caring about of actual pin numbers on both sides
  - add binding for one GPIO line chosen from the above interface that is to allow the nRF9160 to reset the nRF52840
  - update accordingly dts files and board specific code for both board definitions associated with the DK
  - introduce .dtsi files that can be included from dts overlays in order to facilitate the use of the above GPIO interface; modify the overlay in the hci_uart sample to provide an example of use of those files

- Use the multiple board revisions feature to provide support for the new hardware possibilities available in nRF9160 DK starting from v0.14.0.

- Add DT node and .dtsi files that allow using the PCAL6408A I/O expander that is available in nRF9160 DK v0.14.0 or later to interface LEDs and buttons on the DK

anangl committed:
boards: nrf9160dk: Move board control configuration to devicetree

Use devicetree instead of Kconfig to configure the board control
switches in nRF9160 DK:
- add binding for the switches that provide optional signal routings
  on this board
- add binding for the GPIO interface that can be used for communication
  (e.g. UART based) between the nRF9160 and the nRF52840 on the DK,
  and add GPIO mapping for this interface so that its lines can be used
  without caring about of actual pin numbers on both sides
- add binding for one GPIO line chosen from the above interface that is
  to allow the nRF9160 to reset the nRF52840
- update accordingly dts files and board specific code for both board
  definitions associated with the DK
- introduce .dtsi files that can be included from dts overlays in order
  to facilitate the use of the above GPIO interface; modify the overlay
  in the hci_uart sample to provide an example of use of those files

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf9160dk: Add support for newer revisions (0.14.0+) of the DK

Use the multiple board revisions feature to provide support for the new
hardware possibilities available in nRF9160 DK starting from v0.14.0.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

galak reviewed (commented):
Feels like BOARD_REVISION/BOARD_REVISION_STRING need documentation somewhere.

anangl committed:
boards: nrf9160dk: Add DT node for I/O expander available in v0.14.0+

Add devicetree node representing the PCAL6408A I2C-based I/O expander
that is available in nRF9160 DK v0.14.0 or later.
Provide also .dtsi files that can be used in applications to simplify
switching to interfacing onboard LEDs and buttons through this expander
instead of SoC pins.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

tejlmand commented:
> Feels like BOARD_REVISION/BOARD_REVISION_STRING need documentation somewhere.

If you consider the board revision feature itself, then that is documented here:
https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#multiple-board-revisions

If you refer to this commit: https://github.com/zephyrproject-rtos/zephyr/pull/31624/commits/f817ff32e4993ddc733094753ae3193b0119959e
then there is a dedicated PR for that here: https://github.com/zephyrproject-rtos/zephyr/pull/31617

If you mean you would like to see dedicated documentation for those, then it's probably outside the scope of this PR, but I agree that we have too limited documentation for Zephyr / CMake in general.
Updated https://github.com/zephyrproject-rtos/zephyr/issues/9947 with additional content.

anangl commented:
Rebased on #31864.

anangl commented:
Rebased on master after #31864 has been merged.

anangl commented:
Rebased on #31926.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/board.c:
```diff
@@ -1,203 +1,85 @@
-	{ BUTTON1_U12,   IS_ENABLED(CONFIG_BOARD_NRF9160DK_BUTTON0_PHY) },
-	{ BUTTON2_U12,   IS_ENABLED(CONFIG_BOARD_NRF9160DK_BUTTON1_PHY) },
-	{ SWITCH2_U9,    IS_ENABLED(CONFIG_BOARD_NRF9160DK_SWITCH1_PHY) },
+#define GET_CTRL(name, prop, idx) \
```
Might want to call this one GET_CTLR instead, for consistency.

anangl replied:
Of course, that was my intention. I just made a typo here.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,116 @@ Board controller firmware
+
+If you want to route some of the above pins differently or enable any of the
+other available routing options, enable or disable the devicetree node that
+represents the analog switch that provides the giving routing.
```
Did you mean "the given routing" or something like that?

anangl replied:
Yes. Corrected.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,116 @@ Board controller firmware
+on the nRF9160 DK:
+
++------------------------------------+------------------------------+
+| Devicetree node alias              | Analog switch name           |
```
Please use "node label" instead of "node alias", since these are not in `/aliases`.

anangl replied:
Sure, I meant the node label but I used a wrong term.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,116 @@ Board controller firmware
++------------------------------------+------------------------------+
+| Devicetree node alias              | Analog switch name           |
++====================================+==============================+
+| ``vcom0_pins_routing``             | nRF91_UART1 (nRF91_APP1)     |
```
Did you consider naming these to match the analog switch name more closely, e.g. "nrf91_uart1" instead of "vcom0_pins_routing", "nrf91_led1" instead of "led1_pin_routing", etc.? If so, why did you decide against it?

anangl replied:
I thought about it but I was afraid it could be confusing. Especially for the nRF91_UART1 switch which in fact provides routing for `uart0` in the nrf9160dk_nrf9160 board. I wanted to provide more meaningful names as node labels. The nodes themselves are named exactly after the names of analog switches from the [DK documentation](https://infocenter.nordicsemi.com/topic/ug_nrf91_dk/UG/nrf91_DK/board_controller.html). The above node is named `switch-nrf91-uart1` and in the generated `zephyr.dts` file the user will see:
```
		vcom0_pins_routing: switch-nrf91-uart1 {
			compatible = "nordic,nrf9160dk-optional-routing";
			control-gpios = < &gpio1 0xe 0x1 >;
			status = "okay";
		};
```
I think it should be clear enough.
But I'm open to change the naming of those nodes labels if the current one is actually confusing or inconvenient.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,116 @@ Board controller firmware
+| ``nrf_interface_pins_6_8_routing`` | nRF_IF6-8_CTRL (nRF91_COEX)  |
++------------------------------------+------------------------------+
+
+A few helper .dtsi files are provided in dts/ subdirectories for both the
```
Consider using `:zephyr_file:` to link directly to these helpers.

mbolivar-nordic commented on boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml:
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# NOTE: This file is replicated in nrf9160dk_nrf9160 and nrf9160dk_nrf52840.
```
I think it should be possible to use pre_dt_board.cmake from the nrf9160dk_nrf52840 board directory to manipulate DTS_ROOT. By adding the nrf9160dk_nrf9160 board directory to DTS_ROOT, this binding should be available without copy/pasting.

anangl replied:
I'm not sure how to do it properly, as I don't want the other files from `dts/` directory of one board to be available for the other. Only the two files: `nordic,nrf9160dk-nrf52840-interface.yaml` and `nordic,nrf9160dk-nrf52840-reset.yaml` are common for both these boards. Should I move these two files to another directory and add it to DTS_ROOT?

jfischer-no replied:
Is there anything against having only nrf9160dk directory for both boards nrf9160dk_nrf9160 and nrf9160dk_nrf52840?

mbolivar-nordic reviewed (commented):
Nice work!

I need a bit more time with this PR to work through the details, but I have a few minor things from a first pass.

ru-fu commented on boards/arm/nrf9160dk_nrf9160/doc/index.rst:
```diff
@@ -107,6 +107,30 @@ Push buttons and Switches
+   :compact:
+
+Remember to also enable routing for this additional hardware in the firmware
+for :ref:`nrf9160dk_nrf52840` (see the section **Board controller firmware**
```
```suggestion
for :ref:`nrf9160dk_nrf52840` (see the **Board controller firmware** section
```

anangl commented on boards/arm/nrf9160dk_nrf9160/doc/index.rst:
```diff
@@ -107,6 +107,30 @@ Push buttons and Switches
 * SWITCH2 = P0.9
 * BOOT = SW5 = boot/reset
 
+Additional hardware in v0.14.0+
```
Moved.

anangl commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,135 @@ Board controller firmware
+.. note::
+   In nRF9160 DK revisions earlier than v0.14.0, when the above signals are
+   routed to respective VCOMs, LEDs, switches, buttons, or the specific pins
+   of nRF52840, they are not simultaneously available on the DK connectors.
```
I rephrased this note a bit and placed it earlier in the section (right after the first paragraph).

anangl commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,135 @@ Board controller firmware
+If you want to route some of the above pins differently or enable any of the
+other available routing options, enable or disable the devicetree node that
+represents the analog switch that provides the given routing.
+For example, if you want to enable the optional routing for the nRF9160 pins
```
Moved.

anangl commented on boards/arm/nrf9160dk_nrf52840/doc/index.rst:
```diff
@@ -123,37 +123,135 @@ Board controller firmware
-such as LEDs, switches, and specific nRF52840 SoC pins.
+routing specific pins of the nRF9160 SiP to different components on the DK,
+such as LEDs and buttons, UART interfaces (VCOMx) of the interface MCU, and to
+specific nRF52840 SoC pins.
```
I chose the first option, as nRF52840 SoC pins are something similar to those UART interfaces of the interface MCU, so I think they should be regarded as a DK component. :)

anangl commented:
Rebased on master after #31926 has been merged.

anangl commented:
Rebased to solve a merge conflict.

anangl commented:
Rebased to make CI green.

jfischer-no commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ zephyr_library_sources_ifdef(CONFIG_GPIO_XLNX_AXI   gpio_xlnx_axi.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_NPCX       gpio_npcx.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_EMUL       gpio_emul.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_PSOC6      gpio_psoc6.c)
+zephyr_library_sources_ifdef(CONFIG_GPIO_PCAL6408A  gpio_pcal6408a.c)
```
I suggest to open separate PR for PCAL6408A driver.

anangl replied:
Done. See #32748.

anangl commented:
Rebased on #32748.

mbolivar-nordic reviewed (approved):
Lgtm once unblocked.

anangl commented:
Rebased after #32748 has been merged.

ru-fu reviewed (approved):
Docs look good.
